oak-run/src/main/java/org/apache/jackrabbit/oak/run/Downloader.java

```diff
@@ -0,0 +1,263 @@
+                        }
+                    }
+                }).collect(Collectors.toList());
+
+        return new DownloadReport(
+                itemResponses.stream().filter(r -> !r.failed).count(),
+                itemResponses.stream().filter(r -> r.failed).count(),
+                itemResponses.stream().filter(r -> !r.failed).mapToLong(r -> r.size).sum()
+        );
+    }
+
+    @Override
+    public void close() throws IOException {
+        executorService.shutdown();
+    }
+
+    private class DownloaderWorker implements Callable<ItemResponse> {
+
+        private final Item item;
+
+        public DownloaderWorker(Item item) {
+            this.item = item;
+        }
+
+        @Override
+        public ItemResponse call() throws Exception {
+            long t0 = System.nanoTime();
+
+            URLConnection sourceUrl = new URL(item.source).openConnection();
+            sourceUrl.setConnectTimeout(Downloader.this.connectTimeoutMs);
+            sourceUrl.setReadTimeout(Downloader.this.readTimeoutMs);
+
+            Path destinationPath = Paths.get(item.destination);
+            Files.createDirectories(destinationPath.getParent());
+            long size;
+            try (ReadableByteChannel byteChannel = Channels.newChannel(sourceUrl.getInputStream());
+                 FileOutputStream outputStream = new FileOutputStream(destinationPath.toFile())) {
+                size = outputStream.getChannel()
+                        .transferFrom(byteChannel, 0, Long.MAX_VALUE);
+            }
+            long elapsed = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - t0);
+            if (slowLogThreshold > 0 && elapsed >= slowLogThreshold) {
+                LOG.warn("{} [{}] downloaded in {} ms", item.source, IOUtils.humanReadableByteCount(size), elapsed);
+            } else {
+                LOG.debug("{} [{}] downloaded in {} ms", item.source, IOUtils.humanReadableByteCount(size), elapsed);
+            }
+            return ItemResponse.success(size);
+        }
+
+        @Override
+        public String toString() {
+            return "DownloaderWorker{" +
+                    "item=" + item +
+                    '}';
+        }
+    }
+
+    private static class RetryingCallable<V> implements Callable<V> {
+        private final Callable<V> callable;
+        private final int retries;
+
+        public RetryingCallable(int retries, Callable<V> callable) {
+            this.retries = retries;
+            this.callable = callable;
+        }
+
+        public V call() {
+            int retried = 0;
+            // Save exceptions that are thrown after each failure, so they can be printed if all retries fail
+            List<Throwable> exceptions = new ArrayList<>();
+
+            // Loop until it doesn't throw an exception or max number of tries is reached
+            while (true) {
+                try {
+                    return callable.call();
+                } catch (Exception e) {
```

Comment:
Maybe retry only on IOException? In principle, any connection problem will raise an IOException. Any other exception type is likely it is a non-recoverable error. 

And among IOException, we could even go further and try to distinguish between recoverable and non-recoverable exceptions. For instance, an exception indicating that a blob was not found should likely not be retried. But this may be hard to do here because we are working at the level of HTTP protocol, so we would have to parse the response headers to know what was the problem.
 
Did you consider using the Azure Java client instead of working with HTTP connections directly? It could offer some useful functionality. I'm just wondering, as I have no experience with the Azure blob store.

Reply:
The retry now happens only for IOExceptions.

I am not familiar with the Azure Java client. I am not sure what kind of functionalities it exposes. From our initial tests, the current implementation improves the copy performance a lot compare to `azcopy`. This is the main goal. If needed, we could try to directly test the Azure Java SDK in a different PR.